pkg/instancetype/infer/handler.go

```diff
@@ -0,0 +1,111 @@
+/*
+ * This file is part of the KubeVirt project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2024 Red Hat, Inc.
+ *
+ */
+package infer
+
+import (
+	"errors"
+
+	virtv1 "kubevirt.io/api/core/v1"
+	api "kubevirt.io/api/instancetype"
+
+	"kubevirt.io/client-go/kubecli"
+	"kubevirt.io/client-go/log"
+
+	instancetypeErrors "kubevirt.io/kubevirt/pkg/instancetype/errors"
+)
+
+const (
+	logVerbosityLevel = 3
+)
+
+type Handler struct {
+	virtClient kubecli.KubevirtClient
+}
+
+func New(virtClient kubecli.KubevirtClient) *Handler {
+	return &Handler{
+		virtClient: virtClient,
+	}
+}
+
+func shouldIgnoreFailure(ignoreFailurePolicy *virtv1.InferFromVolumeFailurePolicy) bool {
+	return ignoreFailurePolicy != nil && *ignoreFailurePolicy == virtv1.IgnoreInferFromVolumeFailure
+}
+
+func (h *Handler) Instancetype(vm *virtv1.VirtualMachine) error {
+	if vm.Spec.Instancetype == nil {
+		return nil
+	}
+	// Leave matcher unchanged when inference is disabled
+	if vm.Spec.Instancetype.InferFromVolume == "" {
+		return nil
+	}
+
+	ignoreFailure := shouldIgnoreFailure(vm.Spec.Instancetype.InferFromVolumeFailurePolicy)
+	defaultName, defaultKind, err := h.fromVolumes(
+		vm, vm.Spec.Instancetype.InferFromVolume, api.DefaultInstancetypeLabel, api.DefaultInstancetypeKindLabel)
+	if err != nil {
+		var ignoreableInferenceErr *instancetypeErrors.IgnoreableInferenceError
+		if errors.As(err, &ignoreableInferenceErr) && ignoreFailure {
+			log.Log.Object(vm).V(logVerbosityLevel).Info("Ignored error during inference of instancetype, clearing matcher.")
+			vm.Spec.Instancetype = nil
+			return nil
+		}
+		return err
+	}
+
+	if ignoreFailure {
+		vm.Spec.Template.Spec.Domain.Memory = nil
```

Comment:
Add TODOs here for working on GitOps friendliness in future?

Reply:
Added to https://github.com/kubevirt/kubevirt/issues/13044